Frameworks/UIKit/_UIGestureCoordinator.mm

```diff
@@ -0,0 +1,457 @@
+
++ (instancetype)singleton {
+    if (!g_UIGestureCoordinator) {
+        g_UIGestureCoordinator = [_UIGestureCoordinator new];
+    }
+
+    return g_UIGestureCoordinator;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _resetAllTrackingGestures = YES;
+        _gesturesToRemove = [NSMutableSet new];
+        _gesturesInPossibleState = [NSMutableSet new];
+        _regularRecognizers = [NSMutableArray new];
+        _dManipRecognizers = [NSMutableArray new];
+    }
+
+    return self;
+}
+
+// test override to provide custom gestures
+- (void)buildTrackingGesturesListWith:(NSMutableArray*)regularRecognizers
+                     PossibleGestures:(NSMutableSet*)gesturesInPossibleState
+                        DManipGesture:(NSMutableArray*)dManipRecognizers {
+    // by default do nothing
+}
+
+// build tracking gesture list given a view, if view is nil, it invokes buildTrackingGesturesListWith override to get
+// custom build list of gesture for testing
+- (void)_buildTrackingGesturesListForView:(UIView*)view {
+    if (view == nil) {
+        if (_resetAllTrackingGestures) {
+            _resetAllTrackingGestures = FALSE;
+            return [self buildTrackingGesturesListWith:_regularRecognizers
+                                      PossibleGestures:_gesturesInPossibleState
+                                         DManipGesture:_dManipRecognizers];
+        }
+    }
+
+    // Otherwise, build up gesture tracking list based on view
+    if (_resetAllTrackingGestures) {
+        _resetAllTrackingGestures = FALSE;
+
+        // Find gesture recognizers in the hierarchy, back-first
+        UIView* curView = view;
+        while (curView != nil) {
+            // Adding enabled gestures on each view into cooresponding tracking gesture list
+            for (UIGestureRecognizer* curgesture in curView->priv->gestures.get()) {
+                if ([curgesture isEnabled]) {
+                    // Separating gestures into regular (Non-DManip) and DManip Gesture
+                    if (![curgesture isKindOfClass:[_UIDMPanGestureRecognizer class]]) {
+                        [_regularRecognizers addObject:curgesture];
+                    } else {
+                        [_dManipRecognizers addObject:curgesture];
+                    }
+
+                    // if gesture is in possible state, Also add it to possible state list
+                    if ([curgesture state] == UIGestureRecognizerStatePossible) {
+                        [_gesturesInPossibleState addObject:curgesture];
+                    }
+                }
+            }
+
+            curView = curView->priv->superview;
+        }
+    }
+}
+
+// main entrance to send touch to gesture pipleline for processing
+- (bool)processGesturesForTouch:(UITouch*)touch event:(UIEvent*)event touchEventName:(SEL)eventName {
+    if (DEBUG_GESTURES) {
+        TraceVerbose(TAG, L"processGesturesForTouch.");
+    }
+
+    [self _buildTrackingGesturesListForView:touch.view];
```

Comment:
fortunately no. _resetAllTrackingGestures is a switch basically guard against rebuid. if _resetAllTrackingGestures is set to TRUE, normally gets set when all finger lifted off screen.  Otherwise, once it sets to FALSE, stay as FALSE until reset.

---
In reply to: [100588999](https://github.com/Microsoft/WinObjC/pull/1909#discussion_r100588999) [](ancestors = 100588999)